File: conformance/kotlin/IntegrationMain.template.kt
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

private val json = Json { ignoreUnknownKeys = true; isLenient = false }

fun main() {
    val cases = listOf(
        """{"id":"a","quantity":2,"tags":["x","y"]}""" to true,
        """{"id":"a","quantity":1}""" to true,
        """{"id":"a","quantity":0}""" to false,
        """{"id":"a","quantity":1,"tags":["x","x"]}""" to false,
        """{"id":"a","quantity":1,"unknown":true}""" to false,
        """{"quantity":1}""" to false,
    )
    var ok = true
    for ((payload, expected) in cases) {
        val actual = try {
            json.decodeFromString<OrderItem>(payload)
            true
        } catch (e: Exception) {
            false
        }
        if (actual != expected) {
            ok = false
            println("MISMATCH: " + payload + " expected=" + expected + " actual=" + actual)
        }
    }

    // Typed construction is unvalidated by design; validate() /
    // toValidatedJson() round-trip through the serializer explicitly.
    val valid = OrderItem(id = "a", quantity = 1)
    try {
        valid.validate()
    } catch (e: Exception) {
        ok = false
        println("MISMATCH: valid constructed instance failed validate(): " + e)
    }
    val validJson = try {
        valid.toValidatedJson()
    } catch (e: Exception) {
        ok = false
        println("MISMATCH: valid constructed instance failed toValidatedJson(): " + e)
        ""
    }
    if (validJson.isNotEmpty() && !validJson.contains("\"quantity\"")) {
        ok = false
        println("MISMATCH: toValidatedJson() output missing quantity: " + validJson)
    }

    val invalid = OrderItem(id = "a", quantity = 0)
    val invalidThrew = try {
        invalid.validate()
        false
    } catch (e: Exception) {
        true
    }
    if (!invalidThrew) {
        ok = false
        println("MISMATCH: invalid constructed instance passed validate()")
    }
    val invalidJsonThrew = try {
        invalid.toValidatedJson()
        false
    } catch (e: Exception) {
        true
    }
    if (!invalidJsonThrew) {
        ok = false
        println("MISMATCH: invalid constructed instance passed toValidatedJson()")
    }

    println(if (ok) "PASS" else "FAIL")
}

File: conformance/rust/integration-main.template.rs
mod jst_helpers;
mod order_item;

use order_item::OrderItem;

fn main() {
    let cases: &[(&str, bool)] = &[
        (r#"{"id":"a","quantity":2,"tags":["x","y"]}"#, true),
        (r#"{"id":"a","quantity":1}"#, true),
        (r#"{"id":"a","quantity":0}"#, false),
        (r#"{"id":"a","quantity":1,"tags":["x","x"]}"#, false),
        (r#"{"id":"a","quantity":1,"unknown":true}"#, false),
        (r#"{"quantity":1}"#, false),
    ];
    let mut ok = true;
    for (payload, expected) in cases {
        let actual = serde_json::from_str::<OrderItem>(payload).is_ok();
        if actual != *expected {
            ok = false;
            println!("MISMATCH: {payload} expected={expected} actual={actual}");
        }
    }

    // Direct construction (fields are pub), mutation, and re-validation.
    let mut item = OrderItem {
        id: "a".into(),
        quantity: serde_json::Number::from(2),
        tags: None,
    };
    if item.validate().is_err() {
        ok = false;
        println!("MISMATCH: constructed valid item failed validate()");
    }
    item.quantity = serde_json::Number::from(0);
    if item.validate().is_ok() {
        ok = false;
        println!("MISMATCH: quantity=0 passed validate()");
    }
    if item.to_validated_json().is_ok() {
        ok = false;
        println!("MISMATCH: quantity=0 passed to_validated_json()");
    }
    item.quantity = serde_json::Number::from(2);
    if item.to_validated_json().is_err() {
        ok = false;
        println!("MISMATCH: restored item failed to_validated_json()");
    }

    println!("{}", if ok { "PASS" } else { "FAIL" });
}
